nexus/reconfigurator/execution/src/external_networking.rs

```diff
@@ -0,0 +1,1202 @@
+        ));
+
+        let kind = z.zone_type.kind();
+        ensure_external_service_ip(
+            opctx,
+            datastore,
+            kind,
+            z.id,
+            external_ip,
+            &log,
+        )
+        .await?;
+        ensure_service_nic(opctx, datastore, kind, z.id, nic, &log).await?;
+    }
+
+    Ok(())
+}
+
+pub(crate) async fn ensure_zone_external_networking_deallocated(
+    opctx: &OpContext,
+    datastore: &DataStore,
+    zones_to_deallocate: impl Iterator<Item = &BlueprintZoneConfig>,
+) -> anyhow::Result<()> {
+    for z in zones_to_deallocate {
+        let Some((external_ip, nic)) = z.zone_type.external_networking() else {
+            continue;
+        };
+
+        let kind = z.zone_type.kind();
+        let log = opctx.log.new(slog::o!(
+            "action" => "deallocate-external-networking",
+            "zone_kind" => z.zone_type.kind().to_string(),
+            "zone_id" => z.id.to_string(),
+            "ip" => format!("{external_ip:?}"),
+            "nic" => format!("{nic:?}"),
+        ));
+
+        let deleted_ip = datastore
+            .deallocate_external_ip(opctx, external_ip.id().into_untyped_uuid())
+            .await
+            .with_context(|| {
+                format!(
+                    "failed to delete external IP {external_ip:?} \
+                     for {kind} zone {}",
+                    z.id
+                )
+            })?;
+        if deleted_ip {
+            info!(log, "successfully deleted Omicron zone external IP");
+        } else {
+            debug!(log, "Omicron zone external IP already deleted");
+        }
+
+        let deleted_nic = datastore
+            .service_delete_network_interface(
+                opctx,
+                z.id.into_untyped_uuid(),
+                nic.id,
+            )
+            .await
+            .with_context(|| {
+                format!(
+                    "failed to delete service VNIC {nic:?} for {kind} zone {}",
+                    z.id
+                )
+            })?;
+        if deleted_nic {
+            info!(log, "successfully deleted Omicron zone vNIC");
+        } else {
+            debug!(log, "Omicron zone vNIC already deleted");
+        }
+    }
+
+    Ok(())
+}
+
+// Helper function to determine whether a given external IP address is
+// already allocated to a specific service zone.
+async fn is_external_ip_already_allocated(
+    opctx: &OpContext,
+    datastore: &DataStore,
+    zone_kind: ZoneKind,
+    zone_id: OmicronZoneUuid,
+    external_ip: OmicronZoneExternalIp,
+    log: &Logger,
+) -> anyhow::Result<bool> {
+    // localhost is used by many components in the test suite.  We can't use
+    // the normal path because normally a given external IP must only be
+    // used once.  Just treat localhost in the test suite as though it's
+    // already allocated.  We do the same in is_nic_already_allocated().
+    if cfg!(test) && external_ip.ip().is_loopback() {
```

Comment:
I'm a little hesitant to say that's the reason this was put here, but yes, I think that's the reason it will stay here - the tests spin up a bunch of things that all need IPv6 addresses that we can talk to, which means they all have to use `::1`. (I'd claim the reason this was put here was kinda the inverse - all the tests use `::1`, we added the executor and discovered it breaks on IP reuse, so added this "crime" to avoid going back to the tests and figuring out a way to rework them to use something other than `::1`, which I'm not even sure is possible since IPv6 doesn't have a localhost range.)